Osf exercise three

RobertoEhlertOSF opened the pull request:
I had some doubts regarding the scheduler.

In the training, we put the cron trigger on the anounymous screen, and then execute the scheduling, so we can see it through a select on crontrigger table.

So I created two classes for scheduling, one that implements Schedulable, so it will be available for schedule through UI, and the Scheduler, a instance class that adds the cron trigger.

Let me know if I did it as expected.

RobertoEhlertOSF committed:
exercise 3 resolution
misses the test classes

RobertoEhlertOSF committed:
finished exercise 3

ana-levinta commented on force-app/main/default/classes/EmailSummaryforAccounts.cls:
```diff
@@ -0,0 +1,29 @@
+public class EmailSummaryforAccounts {
+    
+	public static void sendEmail(string accountName, string emailAddress, double amountOpen, double amountClosedWon, double amountClosedLost ){
+       Messaging.SingleEmailMessage mail = new Messaging.SingleEmailMessage();
+        
+        mail.toaddresses = new String[] {emailAddress};
+            
+		string emailStyle = '<head> <style type="text/css">' +
+			' .fixed_headers th, .fixed_headers td { padding: 5px; text-align: center; } ' +
+			' .fixed_headers td:nth-child(1), .fixed_headers th:nth-child(1) { min-width: 33%; } ' +
+			' .fixed_headers td:nth-child(2), .fixed_headers th:nth-child(2) { min-width: 33%; } ' +
+			' .fixed_headers td:nth-child(3), .fixed_headers th:nth-child(3) { width: 33%; } ' +
+			' .fixed_headers thead { background-color: #333; color: #FDFDFD; } ' + 
+			' .fixed_headers tbody tr:nth-child(even) { background-color: #DDD; } </style> ';
+        
+		string emailHeader = '<body> <table class="fixed_headers"> ' + 
+			' <thead> <tr> <th>Account</th> <th>Stage</th> <th>Amount</th> </tr> </thead>';
+           
+		string emailBody = ' <tbody> ';
+        
+      	   emailBody += '<tr> <td>' + accountName + '</td> <td> Open 		  </td>	<td> $' + amountOpen 	   + '</td> </tr>' + 
+						'<tr> <td>' + accountName + '</td> <td> Closed Won  </td> <td> $' + amountClosedWon  + '</td> </tr>' +
+						'<tr> <td>' + accountName + '</td> <td> Closed Lost </td> <td> $' + amountClosedLost + '</td> </tr>';
+        
+       	emailBody += '</tbody> </table> </body>';
+		mail.setHtmlBody('<html> ' + emailStyle + emailHeader + emailBody + '</html>');       
+		Messaging.sendEmail(new Messaging.SingleEmailMessage[] { mail });
+    }
```
Please indent the code.

ana-levinta commented on force-app/main/default/classes/OpportunitySummaryByEmailBatch.cls:
```diff
@@ -0,0 +1,66 @@
+@isTest
+public class OpportunitySummaryByEmailBatchTest {
```
This appears to be the test class, contained in the OpportunitySummaryByEmailBatch.cls  
Please add the OpportunitySummaryByEmailBatchTest  class in OpportunitySummaryByEmailBatchTest  file and  commit also the OpportunitySummaryByEmailBatch.cls

ana-levinta commented on force-app/main/default/classes/OpportunitySummaryByEmailBatch.cls:
```diff
@@ -0,0 +1,66 @@
+        	OpportunitySummaryByEmailBatch osb = new OpportunitySummaryByEmailBatch();
+       		Id batchId = Database.executeBatch(osb);
+        Test.stopTest();
+        
```
Each test method should have at least one assert.
Please use System.assert to check the expected behavior.

ana-levinta commented on force-app/main/default/classes/ScheduleSummaryAmountEmail.cls:
```diff
@@ -0,0 +1,8 @@
+    global void execute(SchedulableContext ctx) {
+        
+        OpportunitySummaryByEmailBatch osb = new OpportunitySummaryByEmailBatch();
+        database.executeBatch(osb);
```
Please try to write the code as Apex would be case sensitive, even if it's not.

ana-levinta commented on force-app/main/default/classes/ScheduleSummaryAmountEmailTest.cls:
```diff
@@ -0,0 +1,18 @@
+@isTest
+public class ScheduleSummaryAmountEmailTest {
+    
+	public static string CRON_EXP = '0 0 10 ? 1/1 FRI#4 *';
+	
+    @isTest
+    private static void executeTest(){
+       
+        Test.startTest();
+        	ScheduleSummaryAmountEmail sae = new ScheduleSummaryAmountEmail();        
+        	String jobId = System.schedule('ScheduledApexTest', CRON_EXP, sae);
+        Test.stopTest();
+
+		integer jobsScheduled = [SELECT COUNT() FROM CronTrigger WHERE Id = :jobId];
+    	System.assertEquals(1, jobsScheduled, 'expecting one scheduled job');
+    	
+    }
```
Please pay more attention to indentation and how the code is written.

RobertoEhlertOSF committed:
Fixed regarding comments

ana-levinta commented on force-app/main/default/classes/OpportunitySummaryByEmailBatch.cls:
```diff
@@ -0,0 +1,43 @@
+        return Database.getQueryLocator(
+       		'SELECT Id, Name, ' +
+            '(SELECT email FROM Contacts WHERE Is_Primary_Contact__c = true LIMIT 1), ' +
+            '(SELECT Amount, isWon, isClosed FROM Opportunities) FROM Account'
```
Please select the opportunities that were created/updated in the corresponding month (current month).

ana-levinta commented on force-app/main/default/classes/OpportunitySummaryByEmailBatch.cls:
```diff
@@ -0,0 +1,43 @@
+			for(Opportunity opp : acc.Opportunities ){
+				
+				if(opp.isClosed){
+					amountOpen += opp.Amount;
```
amountOpen should be set if the opportunity is not closed.

ana-levinta commented on force-app/main/default/classes/SchedulerSummaryAmountEmail.cls:
```diff
@@ -0,0 +1,8 @@
+    
+    public void ScheduleEmailJob(){
+        ScheduleSummaryAmountEmail sch = new ScheduleSummaryAmountEmail();
+        string cronExp = '0 0 10 ? 1/1 FRI#4 *';
```
The classes should start with a capital letter.

ana-levinta commented on force-app/main/default/classes/ScheduleSummaryAmountEmailTest.cls:
```diff
@@ -0,0 +1,18 @@
+@isTest
+public class ScheduleSummaryAmountEmailTest {
+    
+	public static string CRON_EXP = '0 0 10 ? 1/1 FRI#4 *';
```
The classes should start with a capital letter.

ana-levinta commented on force-app/main/default/classes/ScheduleSummaryAmountEmailTest.cls:
```diff
@@ -0,0 +1,18 @@
+        String jobId = System.schedule('ScheduledApexTest', CRON_EXP, sae);
+        Test.stopTest();
+
+		integer jobsScheduled = [SELECT COUNT() FROM CronTrigger WHERE Id = :jobId];
```
The classes should start with a capital letter.

RobertoEhlertOSF committed:
Fix the initial char on the classes by capital letters
filter oppoortunities by current month
fixed openAmout with only opened opportunities
